spec/oauth/oidc_spec.lua

```diff
@@ -227,6 +228,103 @@ describe('OIDC', function()
       local credentials, _, _, err = oidc:transform_credentials({ access_token = access_token })
       assert(credentials, err)
     end)
+
+    describe('getting client_id from any JWT claim', function()
+
+      before_each(function()
+        stub(ngx_variable, 'available_context', function(context) return context end)
+      end)
+
+      local get_oidc = function(params)
+        local config = {
+          issuer = 'https://example.com/auth/realms/apicast',
+          config = { id_token_signing_alg_values_supported = { 'RS256' } },
+          keys = { somekid = { pem = rsa.pub } }}
+
+        for key,value in pairs(params) do
+          config[key] = value
+        end
+
+        return _M.new(config)
+      end
+
+      it('use liquid template to modify app_id', function()
+        local oidc = get_oidc({
+          client_id = "{{aud}}",
+          client_id_type = "liquid"})
+
+        local access_token = jwt:sign(rsa.private, {
+          header = { typ = 'JWT', alg = 'RS256', kid = 'somekid' },
+          payload = {
+            iss = oidc_config.issuer,
+            aud = 'foo',
+            azp = 'ce3b2e5e',
+            sub = 'someone',
+            exp = ngx.now() + 10,
+          },
+        })
+
+        local credentials, ttl, _, err = oidc:transform_credentials({ access_token = access_token })
+
+        assert(credentials, err)
+
+        assert.same({ app_id  = "foo" }, credentials)
+        assert.equal(10, ttl)
+      end)
+
+      it('use liquid template functions to modify app_id', function()
+        local oidc = get_oidc({
+          client_id = "{{ custom | first}}",
+          client_id_type = "liquid"})
+
+        local access_token = jwt:sign(rsa.private, {
+          header = { typ = 'JWT', alg = 'RS256', kid = 'somekid' },
+          payload = {
+            iss = oidc_config.issuer,
+            aud = 'notused',
+            azp = 'ce3b2e5e',
+            sub = 'someone',
+            custom = {"foo", "bar"},
+            exp = ngx.now() + 10,
+          },
+        })
+
+        local credentials, ttl, _, err = oidc:transform_credentials({ access_token = access_token })
+
+        assert(credentials, err)
+
+        assert.same({ app_id  = "foo" }, credentials)
+        assert.equal(10, ttl)
+      end)
+
+
+      it('use plaintext format to modify app_id', function()
+
+        local oidc = get_oidc({
+          client_id = "aud",
+          client_id_type = "plain"})
+
+        local access_token = jwt:sign(rsa.private, {
```

Comment:
Can we use the same token for the 3 tests? If that's the case, we should consider defining it in the `describe('getting client_id from any JWT claim', function()` block to avoid duplication.